src/html.cr

```diff
@@ -37,9 +37,31 @@ module HTML
   # io.to_s                          # => "Crystal &amp; You"
   # ```
   def self.escape(string : String, io : IO) : Nil
-    string.each_char do |char|
-      io << SUBSTITUTIONS.fetch(char, char)
+    escape(string.to_slice, io)
+  end
+
+  # Same as `escape(String, IO)` but accepts `Bytes` instead of `String`.
+  #
+  # The slice is assumed to be valid UTF-8.
+  def self.escape(string : Bytes, io : IO) : Nil
+    last_copy_at = 0
+    string.each_with_index do |byte, index|
+      str = case byte
+            when '&'  then "&amp;"
+            when '<'  then "&lt;"
+            when '>'  then "&gt;"
+            when '"'  then "&quot;"
+            when '\'' then "&#39;"
+            else
+              nil
+            end
+      if str
+        io.write_string(string[last_copy_at, index &- last_copy_at])
+        last_copy_at = index &+ 1
+        io << str
+      end
     end
+    io.write_string(string[last_copy_at, string.size &- last_copy_at])
```

Comment:
I don't think this method warrants the use of unsafe arithmetic as an optimization measure:

```suggestion
        io.write_string(string[last_copy_at...index])
        last_copy_at = index + 1
        io << str
      end
    end
    io.write_string(string[last_copy_at..])
```

Reply:
Right. LLVM seems to be smart enough to figure out what I explained here. I guess we can leave it then. However, using unsafe addition should improve performance for non-optimized builds (and non-LLVM backends such as the interpreter).
My preference would be to use `&-` explicitly.